what4/test/Abduct.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Main where
+
+import Data.Foldable (forM_)
+import System.IO (FilePath, IOMode(..), openFile, hClose)
+
+import Data.Parameterized.Nonce (newIONonceGenerator)
+import Data.Parameterized.Some (Some(..))
+
+import What4.Config (extendConfig)
+import What4.Expr
+         ( ExprBuilder,  FloatModeRepr(..), newExprBuilder
+         , BoolExpr, IntegerExpr, GroundValue, groundEval
+         , EmptyExprBuilderState(..))
+import What4.Interface
+         ( BaseTypeRepr(..), getConfiguration
+         , freshConstant, safeSymbol, notPred
+         , impliesPred, intLit, intAdd, intLe )
+import What4.Solver
+import What4.Symbol (SolverSymbol(..))
+import What4.Protocol.SMTLib2 as SMT2
+         (assume, sessionWriter, runCheckSat, runGetAbducts, Writer)
+import What4.Protocol.SMTWriter
+         (mkSMTTerm)
+import What4.Protocol.Online
+
+cvc5executable :: FilePath
+cvc5executable = "cvc5"
+
+main :: IO ()
+main = do
+  Some ng <- newIONonceGenerator
+  sym <- newExprBuilder FloatIEEERepr EmptyExprBuilderState ng
+
+  -- This line is necessary for working with cvc5.
+  extendConfig cvc5Options (getConfiguration sym)
+
+  -- The following is satisfiable. Let's get an abduct from the SMT solver,
+  -- Some formula that will make it unsatisfiable (its negation provable)
+  -- not (y >= 0 => (x + y + z) >= 0)
+  -- This query to the SMT solver: (get-abduct A (=> (>= y 0) (>= (+ x y z) 0)))
+  
+  -- First, declare fresh constants for each of the three variables x, y, z.
+  x <- freshConstant sym (safeSymbol "x") BaseIntegerRepr
+  y <- freshConstant sym (safeSymbol "y") BaseIntegerRepr
+  z <- freshConstant sym (safeSymbol "z") BaseIntegerRepr
+
+  -- Next, build up the clause
+  zero <- intLit sym 0                        -- 0
+  pxyz <- intAdd sym x =<< intAdd sym y z -- x + y + z
+  ygte0 <- intLe sym zero y               -- 0 <= y
+  xyzgte0 <- intLe sym zero pxyz          -- 0 <= (x + y + z) 
+  f <- impliesPred sym ygte0 xyzgte0      -- (0 <= y) -> (0 <= (x + y + z))
+
+  -- Prove f (is ~f unsatisfiable?), and otherwise, print countermodel and a formula that will allow me to prove f
+  putStrLn "Offline SMT calls"
+  prove sym f [ ("x", x)
+              , ("y", y)
+              , ("z", z)
+              ]
+
+  putStrLn "\nOnline call to get abduct"
+  putStrLn "Assert hypothesis, and get-abduct on goal"
+  testGetAbductOnline sym [ygte0] xyzgte0
+
+
+testGetAbducts ::
+  ExprBuilder t st fs ->
+  BoolExpr t ->
+  Int ->
+  IO ()
+testGetAbducts sym f n = do
+  -- Print SMT file in /tmp/
+  mirroredOutput <- openFile "/tmp/what4abduct.smt2" ReadWriteMode
```

Comment:
It would be worth replacing the hard-coded reference to `/tmp/...` here with [`withSystemTempFile`](https://hackage.haskell.org/package/temporary-1.3/docs/System-IO-Temp.html#v:withSystemTempFile) from the `temporary` library, as the latter is more portable across different operating systems.